python/shotgun_fields/shotgun_field_delegate.py

```diff
@@ -92,26 +149,201 @@ def _on_before_paint(self, widget, model_index, style_options):
+        """
+        Gets data from the editor widget and stores it in the specified model at the item index.
+
+        :param editor: The editor widget.
+        :type editor: :class:`~PySide.QtGui.QWidget`
+        :param model: The SG model where the data lives.
+        :type model: :class:`~shotgun_utils.shotgun_model.ShotgunModel`
+        :param index: The index of the model to be edited.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+        """
+        src_index = _map_to_source(index)
+        if not src_index or not src_index.isValid():
+            # invalide index, do nothing
+            return
+
+        new_value = editor.get_value()
+
+        cur_value = src_index.data(shotgun_model.ShotgunModel.SG_ASSOCIATED_FIELD_ROLE)
+        if cur_value == new_value:
+            # value didn't change. nothing to do here.
+            return
+
+        bundle = sgtk.platform.current_bundle()
+
+        # special case for image fields
+        if editor._field_name == "image":
+            primary_item = src_index.model().item(src_index.row(), 0)
+            try:
+                if new_value:
+                    # update the value locally in the model
+                    primary_item.setIcon(QtGui.QIcon(new_value))
+                else:
+                    primary_item.setIcon(QtGui.QIcon())
+            except Exception, e:
+                bundle.log_error("Unable to set icon for widget delegate: %s" % (e,))
+
+            return
+
+        successful = src_index.model().setData(
+            src_index,
+            new_value,
+            shotgun_model.ShotgunModel.SG_ASSOCIATED_FIELD_ROLE,
+        )
+
+        if not successful:
+            bundle.log_error(
+                "Unable to set model data for widget delegate: %s, %s" %
+                (self._entity_type, self._field_name)
+            )
+
+    def editorEvent(self, event, model, option, index):
+        """
+        Handles mouse events on the editor.
+
+        :param event: The event that occurred.
+        :type event: :class:`~PySide.QtCore.QEvent`
+        :param model: The SG model where the data lives.
+        :type model: :class:`~shotgun_utils.shotgun_model.ShotgunModel`
+        :param option: Options for rendering the item.
+        :type option: :class:`~PySide.QtQui.QStyleOptionViewItem`
+        :param index: The index of the model to be edited.
+        :type index: :class:`~PySide.QtCore.QModelIndex`
+
+        :return: ``True``, if the event was handled, ``False`` otherwise.
+        :rtype: ``bool``
+        """
+
+        # Forward mouse clicks to the underlying display widget. This only kicks
+        # in if the editor widget isn't displayed or doesn't process a mouse
+        # event for some reason. If you're having trouble with editors
+        # disappearing, it may be because they can't receive focus or aren't
+        # handling a mouse click.
+        if event.type() == QtCore.QEvent.MouseButtonRelease:
```

Comment:
The primary use for this is labels displaying clickable links (entity, multi-entity, etc). By default, they're painted into the view via the delegate, so you can't interact with them. There were some suggestions online how to work around this that seemed really complex. This is a solution @robblau suggested which I tried and it seems to work... and is much simpler! Basically, detect a mouse click (release is all we have really) in the delegate, populate the underlying widget with the data from the index, then forward the event to the widget. The result is a simulation of clicking on the actual widget. That make sense?